internal/controller/postgrescluster/patroni_test.go

```diff
@@ -39,8 +37,192 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crunchydata/postgres-operator/internal/initialize"
+	"github.com/crunchydata/postgres-operator/internal/naming"
+	"github.com/crunchydata/postgres-operator/pkg/apis/postgres-operator.crunchydata.com/v1beta1"
 )
 
+func TestGeneratePatroniLeaderLeaseService(t *testing.T) {
+	env, cc, _ := setupTestEnv(t, ControllerName)
+	t.Cleanup(func() { teardownTestEnv(t, env) })
+
+	reconciler := &Reconciler{Client: cc}
+
+	cluster := &v1beta1.PostgresCluster{}
+	cluster.Namespace = "ns1"
+	cluster.Name = "pg2"
+	cluster.Spec.Port = initialize.Int32(9876)
+
+	alwaysExpect := func(t testing.TB, service *corev1.Service) {
+		assert.Assert(t, marshalMatches(service.TypeMeta, `
+apiVersion: v1
+kind: Service
+		`))
+		assert.Assert(t, marshalMatches(service.ObjectMeta, `
+creationTimestamp: null
+labels:
+  postgres-operator.crunchydata.com/cluster: pg2
+  postgres-operator.crunchydata.com/patroni: pg2-ha
+name: pg2-ha
+namespace: ns1
+ownerReferences:
+- apiVersion: postgres-operator.crunchydata.com/v1beta1
+  blockOwnerDeletion: true
+  controller: true
+  kind: PostgresCluster
+  name: pg2
+  uid: ""
+		`))
+		assert.Assert(t, marshalMatches(service.Spec.Ports, `
+- name: postgres
+  port: 9876
+  protocol: TCP
+  targetPort: postgres
+		`))
+
+		// Always gets a ClusterIP (never None).
+		assert.Equal(t, service.Spec.ClusterIP, "")
+		assert.Assert(t, service.Spec.Selector == nil,
+			"got %v", service.Spec.Selector)
+	}
+
+	t.Run("NoServiceSpec", func(t *testing.T) {
+		service, err := reconciler.generatePatroniLeaderLeaseService(cluster)
+		assert.NilError(t, err)
+		alwaysExpect(t, service)
+
+		// Defaults to ClusterIP.
+		assert.Equal(t, service.Spec.Type, corev1.ServiceTypeClusterIP)
+	})
+
+	t.Run("AnnotationsLabels", func(t *testing.T) {
+		cluster := cluster.DeepCopy()
+		cluster.Spec.Metadata = &v1beta1.Metadata{
+			Annotations: map[string]string{"a": "v1"},
+			Labels:      map[string]string{"b": "v2"},
+		}
+
+		service, err := reconciler.generatePatroniLeaderLeaseService(cluster)
+		assert.NilError(t, err)
+
+		// Annotations present in the metadata.
+		assert.DeepEqual(t, service.ObjectMeta.Annotations, map[string]string{
+			"a": "v1",
+		})
+
+		// Labels present in the metadata.
+		assert.DeepEqual(t, service.ObjectMeta.Labels, map[string]string{
+			"b": "v2",
+			"postgres-operator.crunchydata.com/cluster": "pg2",
+			"postgres-operator.crunchydata.com/patroni": "pg2-ha",
+		})
+
+		// Labels not in the selector.
+		assert.Assert(t, service.Spec.Selector == nil,
+			"got %v", service.Spec.Selector)
+	})
+
+	types := []struct {
+		Type   string
+		Expect func(testing.TB, *corev1.Service)
```

Comment:
Why `testing.TB` here :question:

Reply:
I have a habit of using this interface for helpers. Assertion libraries use a much smaller one so they can test themselves and stay compatible with different versions of Go: 1. [gotest.tools](https://pkg.go.dev/gotest.tools/v3@v3.0.3/assert#TestingT), 2. [testify](https://pkg.go.dev/github.com/stretchr/testify@v1.7.0/assert#TestingT), 3. [gomega](https://pkg.go.dev/github.com/onsi/gomega@v1.15.0/types#GomegaTestingT)

Reply:
hm good to know :+1: